cli/popper/runner_slurm.py

```diff
@@ -0,0 +1,70 @@
+import os
+from subprocess import PIPE, Popen, STDOUT, SubprocessError
+
+import docker
+
+from popper import utils as pu
+from popper import scm
+from popper.cli import log as log
+from popper.runner import StepRunner as StepRunner
+from popper.runner_host import DockerRunner as HostDockerRunner
+from popper.runner_host import HostRunner
+
+
+class SlurmRunner(StepRunner):
+
+    spawned_processes = set()
+
+    def __init__(self, config):
+        super(SlurmRunner, self).__init__(config)
+
+    def __exit__(self, exc_type, exc, traceback):
+        SlurmRunner.spawned_processes = set()
+
+    def exec_srun_cmd(self, cmd, env=None):
+        cmd.insert(0, 'srun')
+        ecode = pu.exec_cmd(
+            cmd, env, self.config.workspace_dir, SlurmRunner.spawned_processes)
+        return ecode
+
+    def stop_srun_cmd(self):
+        for p in SlurmRunner.spawned_processes:
+            log.info(f'Stopping proces {p.pid}')
+            p.kill()
+
+
+class DockerRunner(SlurmRunner, HostDockerRunner):
+    """Runs steps in docker."""
+
+    def __init__(self, config):
+        super(DockerRunner, self).__init__(config)
+
+    def run(self, step):
+        """Execute the given step in docker."""
+        cid = pu.sanitized_name(step['name'], self.config.wid)
+
+        container = HostDockerRunner.find_container(cid)
+
+        if container and not self.config.reuse and not self.config.dry_run:
+            container.remove(force=True)
+
+        container = HostDockerRunner.create_container(step, self.config)
+        log.info(f'[{step["name"]}] srun docker start')
+
+        if self.config.dry_run:
+            return 0
+
+        HostDockerRunner.spawned_containers.append(container)
+        ecode = self.start_container(cid)
```

Comment:
in a slurm setup, the above runs on the machine where `popper run` is being executed (i.e. the frontend node). But the goal is to run this on multiple nodes, so all this needs to be executed by slurm. One thing we can do is create a script and instruct slurm to execute it, but it looks like `srun` doesn't support this, so we'll need to use `sbatch`.

Reply:
another thing that this brings up is how to expose options for slurm jobs. This can be done at the level of the step:

```yaml
- id: 'mpi-step'
  uses: 'docker://an-mpi-image'
  options:
    slurm:
      cpus-per-task: 4
      nodes: 64
      ...
```

and the above would be communicated to `sbatch` in the script as explained [here](https://help.rc.ufl.edu/doc/Sample_SLURM_Scripts). The other alternative is to have it in the config file like:

```yaml
resource_manager:
  name: slurm
  options:
    mpi-step:
      cpus-per-task: 4
      nodes: 64
      ...
```

Reply:
I think having the options in the `settings.yml` is a good idea.